cylc/flow/data_store_mgr.py

```diff
@@ -766,100 +839,122 @@ def update_data_structure(self, updated_nodes=None):
-        if new_points or prune_points:
-            # Pruned and/or additional elements require
-            # state/status recalculation, and ID ref updates.
+            self.deltas[TASK_PROXIES].pruned.append(tp_id)
+            self.schd.job_pool.remove_task_jobs(tp_id)
+            self.deltas[EDGES].pruned.extend(node.edges)
+            parent_ids.add(node.first_parent)
+
+        prune_ids = set()
+        checked_ids = set()
+        while parent_ids:
+            for fp_id in parent_ids:
+                self._family_ascent_point_prune(
+                    fp_id, node_ids, parent_ids, checked_ids, prune_ids)
+                break
+        if prune_ids:
+            self.deltas[FAMILY_PROXIES].pruned.extend(prune_ids)
             self.updates_pending = True
 
-    def prune_points(self, point_strings):
-        """Remove old nodes and edges by cycle point.
+    def _family_ascent_point_prune(
+            self, fp_id, node_ids, parent_ids, checked_ids, prune_ids):
+        """Find and prune family recursively checking child families.
 
-        Args:
-            point_strings (iterable):
-                Iterable of valid cycle point strings.
+        Recursively map out child families to the bottom from the origin
+        family. The work back up to origin checking these families are active.
 
         """
-        flow_data = self.data[self.workflow_id]
-        if not point_strings:
+        fp_data = self.data[self.workflow_id][FAMILY_PROXIES]
+        if fp_id in fp_data and fp_id not in self.updated[FAMILY_PROXIES]:
+            fam_node = fp_data[fp_id]
+            # Gather child families, then check/update recursively
+            child_fam_nodes = [
+                n_id
+                for n_id in fam_node.child_families
+                if n_id not in checked_ids
+            ]
+            for child_id in child_fam_nodes:
+                self._family_ascent_point_prune(
+                    child_id, node_ids, parent_ids, checked_ids, prune_ids)
+            if [
+                    child_id
+                    for child_id in fam_node.child_tasks
+                    if child_id not in node_ids
+            ] or [
+                    child_id
+                    for child_id in fam_node.child_families
+                    if child_id not in prune_ids
+            ]:
+                self.state_update_families.add(fp_id)
+            else:
+                if fam_node.first_parent:
+                    parent_ids.add(fam_node.first_parent)
+                prune_ids.add(fp_id)
+        checked_ids.add(fp_id)
+        if fp_id in parent_ids:
+            parent_ids.remove(fp_id)
+
+    def update_dynamic_elements(self, updated_nodes=None):
+        """Update data elements containing dynamic/live fields."""
+        # If no tasks are given update all
+        if updated_nodes is None:
+            updated_nodes = self.schd.pool.get_all_tasks()
+        elif not updated_nodes:
```

Comment:
Should it be an `if` instead? In case the `updated_nodes = self.schd.pool.get_all_tasks()` makes `updated_nodes` to be `False`?

Reply:
Yeah could be `if`.. The assumption is `self.schd.pool.get_all_tasks()` is never empty thereafter, but even if it is, the following loop will do nothing.